File: src/main/kotlin/commands/Commands.kt
package commands

import models.CustomSubject
import org.casbin.adapter.JDBCAdapter
import org.casbin.jcasbin.main.Enforcer
import org.casbin.jcasbin.persist.file_adapter.FilteredAdapter
import utils.getEnforcer
import utils.getModel
import utils.getPostgresAdapter
import java.util.*
import kotlin.system.measureTimeMillis

fun dbSetup(args: List<String>) {
    val enforcer = getEnforcer(null, null)
    enforcer.addNamedGroupingPolicies(
        "g", listOf(
            listOf("alice", "admin", "domain1"),
            listOf("bob", "admin", "domain2"),
            listOf("ale", "admin", "domain1"),
        )
    )
    enforcer.addNamedGroupingPolicies(
        "g2", listOf(
            listOf("content", "root"),
            listOf("course", "content"),
            listOf("exam", "content"),
            listOf("course1", "course"),
            listOf("course2", "course"),
            listOf("exam1", "exam"),
            listOf("exam2", "exam")
        )
    )
    enforcer.addPolicies(
        listOf(
            listOf("admin", "*", "data1", "read"),
            listOf("admin", "*", "data1", "write"),
            listOf("admin", "domain2", "data2", "read"),
            listOf("admin", "domain2", "data2", "write"),
            listOf("admin", "*", "content", "*"),
        )
    )
    enforcer.savePolicy()
    println("db setup completed")
}

fun dbSetupLarge(args: List<String>) {
    val enforcer = getEnforcer(null, null)
    val names = getNames()
    val roles = getRoles()
    val domains = getDomains()
    val pattern = """(?i)\d""".toRegex()
    val resources = getResources().map {listOf(it, it.take(pattern.find(it)?.range?.first ?: it.length))}
    val userRoles =
        generateSequence {
            listOf(
                names.shuffled().first(),
                roles.shuffled().first(),
                domains.shuffled().first()
            )
        }.take(1000).toList()
    val types = listOf("course", "exam", "quiz", "lab")
    val genType = {
        val t = types.shuffled().first()
        listOf(t + UUID.randomUUID(), t)
    }
    val randomResources = generateSequence { genType() }.take(10000).toList()
    val resourceRoles = listOf(
        listOf("content", "root"),
        listOf("course", "content"),
        listOf("exam", "content"),
        listOf("lab", "content"),
    )
    val actions = listOf("read", "edit", "consume", "share")
    val policies = generateSequence {
        listOf(
            (names.plus("manager")).shuffled().first(),
            domains.shuffled().first(),
            resources.shuffled().first()[0],
            actions.shuffled().first()
        )
    }.take(1000).toList()
    enforcer.addNamedGroupingPolicies(
        "g", userRoles
    )
    enforcer.addNamedGroupingPolicies(
        "g2", resourceRoles + resources + randomResources
    )
    enforcer.addPolicies(policies)
    enforcer.savePolicy()
    println("db setup completed")
}

fun benchmark(args: List<String>) {
    val start = System.currentTimeMillis()

//    val enforcer = getEnforcer(null, null)
    println("enforcer init took ${System.currentTimeMillis() - start} ms")
    val model = getModel()
    val adapter = getPostgresAdapter() as JDBCAdapter

    val names = getNames()
    val actions = listOf("read", "edit", "consume", "share")
    val objects = getResources()
    val domains = getDomains()
    val nPolicies = if (args.size == 1) args[0].toInt() else 100
    val timeInMillis = measureTimeMillis {
        for (k in 0..nPolicies) {
            val name = names.shuffled().first()
            val domain = domains.shuffled().first()
            val obj = objects.shuffled().first()
            val act = actions.shuffled().first()
            val isAdmin = listOf(true, false).shuffled().first()

            val sub = CustomSubject(name = name, isAdmin = isAdmin) // the user that wants to access a resource.

            val policyFilter = FilteredAdapter.Filter()
            policyFilter.g = arrayOf("", "")
            policyFilter.p = arrayOf(
                name, "", "", ""
            )
            val enforcer = Enforcer(model)
            enforcer.adapter = adapter
            enforcer.loadFilteredPolicy(policyFilter)
            val result = enforcer.enforce(sub, domain, obj, act)
//            println(result)
        }
    }
    println("Evaluation of $nPolicies policies took $timeInMillis ms")
}

fun checkPolicy(args: List<String>) {
    if (args.size < 4) {
        println("not enough arguments $args")
        return
    }

    val enforcer = getEnforcer(null, null)
    val subject = args[0]
    val domain = args[1]
    val obj = args[2]
    val action = args[3]
    val sub = CustomSubject(name = subject, isAdmin = false) // the user that wants to access a resource.
    println("Checking policy $args. Result: ${enforcer.enforce(sub, domain, obj, action)}")
}

File: golang/cmd/casbinSample/commands.go
package utils

import (
	"bufio"
	"fmt"
	"github.com/apex/log"
	"github.com/casbin/casbin/v2"
	"github.com/spf13/cobra"
	"golang/cmd/utils"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type CommandsManager struct {
	Enforcer *casbin.CachedEnforcer
}

func NewManager() CommandsManager {
	start := time.Now().UnixMilli()
	enforcer, err := utils.GetEnforcer()
	log.WithField("timeSpent", time.Now().UnixMilli()-start).Info("enforcer init took {timeSpent}")

	if err != nil {
		panic(err)
	}
	return CommandsManager{
		Enforcer: enforcer,
	}
}

func (m CommandsManager) SetupDB() *cobra.Command {
	run := func(cmd *cobra.Command, args []string) error {
		m.setupDB()
		return nil
	}

	cmd := &cobra.Command{
		Use:  "setup-db",
		RunE: run,
	}
	return cmd
}

func (m CommandsManager) CheckPolicy() *cobra.Command {
	run := func(cmd *cobra.Command, args []string) error {
		m.checkPolicy(args)
		return nil
	}

	cmd := &cobra.Command{
		Use:  "check-policy",
		RunE: run,
	}
	return cmd
}

func (m CommandsManager) Benchmark() *cobra.Command {
	run := func(cmd *cobra.Command, args []string) error {
		m.benchmark(args)
		return nil
	}

	cmd := &cobra.Command{
		Use:  "benchmark",
		RunE: run,
	}
	return cmd
}

func (m CommandsManager) Interactive() *cobra.Command {
	run := func(cmd *cobra.Command, args []string) error {
		m.interactive(args)
		return nil
	}

	cmd := &cobra.Command{
		Use:  "interactive",
		RunE: run,
	}
	return cmd
}

func (m CommandsManager) setupDB() {
	enforcer := m.Enforcer

	userRoles := [][]string{
		{"alice", "admin", "domain1"},
		{"bob", "admin", "domain2"},
		{"ale", "admin", "domain1"},
	}
	_, _ = enforcer.AddNamedGroupingPolicies("g", userRoles)

	resourceRoles := [][]string{
		{"content", "root"},
		{"course", "content"},
		{"exam", "content"},
		{"course1", "course"},
		{"course2", "course"},
		{"exam1", "exam"},
		{"exam2", "exam"},
	}
	_, _ = enforcer.AddNamedGroupingPolicies("g2", resourceRoles)

	policies := [][]string{
		{"admin", "*", "data1", "read"},
		{"admin", "*", "data1", "write"},
		{"admin", "domain2", "data2", "read"},
		{"admin", "domain2", "data2", "write"},
		{"admin", "*", "content", "*"},
	}
	_, _ = enforcer.AddPolicies(policies)
	_ = enforcer.SavePolicy()
}

func (m CommandsManager) checkPolicy(policy []string) {
	enforcer := m.Enforcer

	subject := policy[0]
	domain := policy[1]
	obj := policy[2]
	action := policy[3]
	sub := CustomSubject{Name: subject, IsAdmin: false} // the user that wants to access a resource.
	result, err := enforcer.Enforce(sub, domain, obj, action)
	if err != nil {
		log.WithError(err).Fatal("Error")
	}
	log.WithField("policy", policy).WithField("result", result).
		Info("Checking policy {policy}. Result: {result}")
}

func (m CommandsManager) benchmark(args []string) {
	enforcer := m.Enforcer

	names := utils.GetNames()
	actions := []string{"read", "edit", "consume", "share"}
	objects := utils.GetResources()
	domains := utils.GetDomains()
	nPolicies := 100
	if len(args) > 0 {
		nPolicies, _ = strconv.Atoi(args[0])
	}

	start := time.Now().UnixMilli()
	rand.Seed(time.Now().UnixNano())
	for i := 1; i <= nPolicies; i++ {
		name := utils.RandomItem(names)
		domain := utils.RandomItem(domains)
		obj := utils.RandomItem(objects)
		act := utils.RandomItem(actions)
		isAdmin := false
		sub := CustomSubject{Name: name, IsAdmin: isAdmin}
		_, err := enforcer.Enforce(sub, domain, obj, act)
		if err != nil {
			log.WithError(err).Fatal("Error")
		}
	}
	log.WithField("nPolicies", nPolicies).
		WithField("timeSpent", time.Now().UnixMilli()-start).
		Info("Computing {nPolicies} policies took {timeSpent} ms")
}

func (m CommandsManager) interactive(args []string) {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter command: ")
	input, _ := reader.ReadString('\n')
	for input != "exit" {
		i := strings.Split(strings.TrimSuffix(input, "\n"), " ")
		command := i[0]
		args = i[1:]

		switch command {
		case "check":
			m.checkPolicy(args)
		case "setup-db":
			m.setupDB()
		case "benchmark":
			m.benchmark(args)
		default:
			break
		}
		fmt.Print("Enter command: ")
		input, _ = reader.ReadString('\n')
	}
}
